### ISSUE TITLE
False positives and negatives of NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE

### ISSUE THREAD

**Issue by davidburstromspotify** [spotbugs/spotbugs#782]
There are some false positives and negatives of NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE related to various combinations of custom nullables and null checks. One of them is especially nefarious, as it relies solely on the formatting of the source code. See the example project which reproduces the issues.

[Example project](https://github.com/davidburstromspotify/spotbugs-checknotnull-bug/blob/master/src/main/java/Example.java)

**Comment by bamapookie** [comment]
I am getting this as well on repeated calls to a @Nullable getter.  Null has already been checked for prior to use and the use gives this false positive.

**Comment by davidburstromspotify** [comment]
@bamapookie That is not a false positive. The `@Nullable` annotation has no indication of invariance, meaning that for whichever reason, the getter could actually return `null` the second time around, e.g. due to the logic of the getter or through concurrent updates. Although SpotBugs could analyse the getter to check if the returned value is literal or final, this does not apply for invocations on interfaces where the concrete implementation is unknown. If invariance is determined, it would also make library version bumps trickier. If the implementation of a library method changes to be variant, you'd need to deal with this at all your call sites.

The workaround is to assign the return value to a local variable.

**Comment by bamapookie** [comment]
Fair enough.  And assigning the local variable is exactly what I did. :)

Thanks.

**Comment by olegshtch** [comment]
> The `@Nullable` annotation has no indication of invariance, meaning that for whichever reason, the getter could actually return `null` the second time around, e.g. due to the logic of the getter or through concurrent updates.

Checker framework uses `@Pure` annotation to declare invariance.
